README.md

```diff
@@ -1,16 +1,644 @@
+
+## 5.1 Promises
+
+## 5.1.1 Fields
+
+| Field         | Type         | Description                                                                           | Required | Default  |
+|---------------|--------------|---------------------------------------------------------------------------------------|----------|----------|
+| `promised`    | `CID or "/"` | The Invocation being referenced                                                       | Yes      |          |
+| `actionlabel` | `String`     | The action's label. If the actions were implicit (`"run": "*"`), the the CID is used  | Yes      |          |
+
+The above table MUST be serialized as a tuple. In JSON, this SHOULD be represented as an array containing the values (but not keys) sequenced in the order they appear in the table.
+
+## 5.1.2 IPLD Schema
+
+``` ipldsch
+type Promise struct {
+  promised    Target 
+  actionlabel String                   -- The label inside the invocation
+} representation tuple
+
+type Target enum {
+  | Local (rename "/")
+  | Remote(CID)
+}
+```
+
+## 5.1.3 JSON Examples
+
+``` js
+["/", "some-label"]
+["QmYW8Z58V1v8R25USVPUuFHtU7nGouApdGTk3vRPXmVHPR", "some-label"]
+["QmYW8Z58V1v8R25USVPUuFHtU7nGouApdGTk3vRPXmVHPR", "QmeURvKzGm85Ekt1t1wDJ3niHAXuhexi281N2ig311pLZv"]
+```
+
+## 5.2 Pipelining
+
+Pipelining uses promises as inputs to determine the required dataflow graph. The following examples both express the following dataflow graph:
+
+```
+              ┌────────────────────────────┐
+              │                            │
+              │ dns://example.com?TYPE=TXT │
+              │        crud/update         │
+              │                            │
+              └───────┬──────────┬─────────┘
+                      │          │
+                      │          │
+┌─────────────────────▼───┐  ┌───▼────────────────────────┐
+│                         │  │                            │
+│ mailto:alice@exaple.com │  │ mailto://alice@example.com │
+│         msg/send        │  │          msg/send          │
+│     bob@example.com     │  │      carol@exmaple.com     │
+│                         │  │                            │
+└─────────────────────┬───┘  └───┬────────────────────────┘
+                      │          │
+                      │          │
+             ┌────────▼──────────▼────────┐
+             │                            │
+             │ https://example.com/events │
+             │         crud/create        │
+             │                            │
+             └────────────────────────────┘
+```
+
+To keep things simple, we expect the shape of the return data of all of these steps to include:
+
+``` json
+{"http": {"body": "some text"}}
+```
+
+#### 5.2.1 Batched 
+
+ ``` json
+{
+  "ucan/invoke": "QmYW8Z58V1v8R25USVPUuFHtU7nGouApdGTk3vRPXmVHPR",
+  "v": "0.1.0",
+  "nnc": "abcdef",
+  "ext": null,
+  "run": {
+    "update-dns" : {
+      "using": "dns://example.com?TYPE=TXT":
```

Comment:
```suggestion
      "with": "dns://example.com?TYPE=TXT":
```

I'd personally stick to the same naming as in UCANs to keep it simple & familiar

Reply:
Fair! We're going to remove those in UCAN 0.10 though, right? I also heard recently that `with` can cause problems for people because it's a keyword in a lot of languages.

Happy to switch this to `with` if you still think it's better 👍 Lemme know

Reply:
I switched this to `with` for now, since you're right that is the current term that we use 💯